### PR TITLE
fcm make: load previous ctx: support relocate

### LINKED ISSUE #185: fcm make: unable to relocate dest due to absolute paths in ctx.gz
We have shared directories on both our front end and HPC machines where we'd like to keep UM prebuilds. Because of the way permissions are set up it's difficult to write to these with a normal fcm make.The preferred approach would be to create the prebuild as a regular user and then move the prebuild directories to the shared areas.

However moving prebuilds doesn't seem to work. A make using the prebuild still tries to access files from the original location. The problem seems to be the full file paths for the extract and cache directories in .fcm_make/ctx.gz.

My perl isn't up to properly understanding the storable data structure, but if I do a brute force string replace of the path names in this file (and fix the string length headers) then I get something that works. (Our HPC system is down for maintenance today so I got further into this than I really intended)

```python
import gzip
f = gzip.open('ctx.gz', 'rb')
data = f.read()
f.close()

old_path = '/home/user/prebuild'
new_path = '/home/shared/prebuild'

newdata = data.replace(old_path, new_path)
# Now fix up all the string lengths
delta = len(new_path) - len(old_path)
# Strings don't support item assignment so use an array
b = bytearray(newdata)
start = 0
pos = newdata.find(new_path,start)
while pos > 0:
    b[pos-1] += delta
    start = pos + 1
    pos = newdata.find(new_path,start)

f = gzip.open('newctx.gz', 'wb')
f.write(str(b))
f.close()
```

Is there some more elegant way around this? I'm using 2015.03.0. I was hopeful when I saw the comment in the release notes "This should make it easier to relocate a build" but it hasn't helped this problem.

### PR DESCRIPTION
**Pull request by matthewrmshin**
If an old make context's `dest` setting does not agree with the location of where it is loaded, the loader will attempt to update any relevant path related settings that are paths under the old `dest` setting.

Fix #185.

@benfitzpatrick please review.

### CODE REVIEW

File: lib/FCM/System/Make/Build.pm [hunk 1]
```diff
@@ -304,6 +305,45 @@ sub _ctx {
...
+sub _ctx_load_hook {
+    my ($attrib_ref, $old_m_ctx, $old_ctx, $old_m_dest, $old_dest) = @_;
+    my $path_mod_func = sub {
+        my ($get_func, $set_func) = @_;
+        my $path = $get_func->();
+        if (!defined($path)) {
+            return;
+        }
+        my $rel_path = abs2rel($path, $old_m_dest);
+        if (index($rel_path, '..') != 0) {
```
Comment (benfitzpatrick) [inline lib/FCM/System/Make/Build.pm]: Could have a directory name with `..`

**Comment by benfitzpatrick** [comment]
I think the new test passes with current master - could you confirm that?

File: lib/FCM/System/Make/Build.pm (hunk 1 above)
Comment (matthewrmshin) [inline lib/FCM/System/Make/Build.pm]: Yes, but that kind of behaviour is asking for it!

**Comment by benfitzpatrick** [comment]
It's a strange behaviour in the fcm test battery lib/bash/test_header, I think. It can set `FCM_HOME` regardless of where the parent `fcm test-battery` command is (/opt/fcm/bin/fcm test-battery /some/path/to/other/fcm/t/fcm-make/45-x.t uses /some/path/to/other/fcm).

**Comment by matthewrmshin** [comment]
> I think the new test passes with current master - could you confirm that?

It fails on current master.

> It's a strange behaviour in the fcm test battery `lib/bash/test_header`, I think.

Please feel free to raise a separate issue about this.

**Pull request merged**